rkt/fetch.go

```diff
@@ -47,9 +59,106 @@ func init() {
 	cmdFetch.Flags().Var((*appAsc)(&rktApps), "signature", "local signature file to use in validating the preceding image")
 	cmdFetch.Flags().BoolVar(&flagStoreOnly, "store-only", false, "use only available images in the store (do not discover or download from remote URLs)")
 	cmdFetch.Flags().BoolVar(&flagNoStore, "no-store", false, "fetch images ignoring the local store")
+	cmdFetch.Flags().IntVar(&flagP2pDuration, "p2p-duration", 10, "p2p continue service duration (minutes) after download finished")
+}
+
+func p2pFetch(args []string) int {
+	if len(args) < 1 {
+		fmt.Printf("rkt fetch must provide a tottent file.")
+		return 1
+	}
+
+	duration := strconv.Itoa(flagP2pDuration)
+	//start p2p client
+	var cmd *exec.Cmd
+	cmd = exec.Command("./torrent", args[0], duration)
+	if err := cmd.Start(); err != nil {
+		fmt.Printf("start p2p process err:%s\n", err)
+		return 1
+	}
+
+	pid := cmd.Process.Pid
+
+	//handle Ctrl+C signal
+	go func() {
+		signalChan := make(chan os.Signal, 1)
+		signal.Notify(signalChan, os.Interrupt, os.Kill)
+
+		<-signalChan
+		if err := syscall.Kill(pid, 9); err != nil {
+			fmt.Printf("kill p2p client process err:%s\n", err)
+		}
+	}()
+
+	//wait for p2p client start
+	time.Sleep(time.Second * time.Duration(7))
+
+	var (
+		client  *rpc.Client
+		err     error
+		s       *store.Store
+		aciFile *os.File
+		key     string
+	)
+	//connet to p2p client process and get download rate
+	if client, err = rpc.DialHTTP("tcp", "127.0.0.1:1234"); err != nil {
+		fmt.Printf("connet to p2p client err:%s\n", err)
+		return 1
+	}
+	reply := make([]string, 1)
+	if err = client.Call("Download.GetDownloadTotalSize", struct{}{}, &reply); err != nil {
+		fmt.Printf("get download total size err:%s\n", err)
+		return 1
+	}
+	totalSize := reply[0]
+
+	if err = client.Call("Download.GetDownloadFile", struct{}{}, &reply); err != nil {
+		fmt.Printf("get download file err:%s\n", err)
+		return 1
+	}
+	aciImage := reply[0]
+
+	//get rate for download
+	for {
+		if err = client.Call("Download.GetDownloadRate", struct{}{}, &reply); err != nil {
+			fmt.Printf("get download rate err:%s\n", err)
+			return 1
+		}
+		fmt.Printf("\rtotal size:%sKB, downloaded:%s", totalSize, reply)
+		if reply[0] == "100.00%" {
+			break
+		}
+		time.Sleep(time.Second * time.Duration(5))
+	}
+
+	//save aci to rkt store
+	if s, err = store.NewStore(globalFlags.Dir); err != nil {
+		fmt.Printf("open rkt store err:%s\n", err)
+		return 1
+	}
+	if aciFile, err = os.Open(aciImage); err != nil {
+		fmt.Printf("opening ACI file %s err:%s\n", aciImage, err)
+		return 1
+	}
+	if key, err = s.WriteACI(aciFile, true); err != nil {
+		fmt.Printf("write ACI file err:%s\n", err)
+		return 1
+	}
+	fmt.Println(key)
+
+	return 0
 }
 
 func runFetch(cmd *cobra.Command, args []string) (exit int) {
+	//start p2p download if args[0] is a torrent file
+	file := strings.TrimSpace(args[0])
+	if _, err := os.Stat(file); err == nil || os.IsExist(err) {
```

Comment:
This is wrong. os.IsExists checks `err` for calls which tries to create directory entry (file, directory, named pipe or something other) and this entry already exists. This also would be close to working code, if You would try to open file for read access, and then tested if os.IsNotExist(err) would be passed as error.

If You wan't to test if it exists (and if it's file, I'm sure that os.stat and path.ext would also in Your example pass clean for directory with such suffix)  - You should check os.Stat output values (which now You are ignoring) with `output_value.IsRegular()` ( https://golang.org/pkg/os/#FileMode.IsRegular ) but also in this situation You should check if file is accessible by user (in some future maybe rkt will support running as regular user instead of requiring `root` rights).

IMO the easiest way to achieve this is:
* check path extension - if it's ".torrent". if it is:
  * check if path can be opened for reading. if it can be - close file, return p2pFetch...